### PR TITLE
[dynamo] Use polyfill to trace through the attributes of torch.jit.* and lru_cache_wrapper

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128336
* #129117

Earlier we were taking the vt for `obj` and then monkeypatching that `vt.source` to be `obj._torchdynamo_inline`. If one accesses `obj.attr_a`, this would cause problems because Dynamo would then search it in `obj._torchdynamo_inline.attr_a`. This PR makes it more functional, so that we have different vts for obj and `ob._torchdynamo_inline`.

Fixes https://github.com/pytorch/pytorch/issues/93698

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

**Review by jansel — APPROVED** [review]

**Comment by anijain2305** [comment]
Redid the PR and made the tracing of jit functions and lru cache wrappers more functional. So, we now have two vts - one for wrapper, one for inner. Therefore, sources are tracked nicely (earlier we were monkeypatching the outer vt to behave like the inner member).

**Review by yanboliang — APPROVED** [review]

**Review by jansel — APPROVED** [review]

File: torch/_dynamo/variables/builder.py [hunk 1]
```diff
@@ -889,6 +898,14 @@ def build_key_value(i, k, v):
...
             return TorchCtxManagerClassVariable(value, source=self.source)
+        elif inspect.getattr_static(value, "__script_if_tracing_wrapper", False):
+            self.install_guards(GuardBuilder.TYPE_MATCH)
+            return WrapperUserFunctionVariable(
+                value, "__original_fn", source=self.source
+            )
+        elif is_lru_cache_wrapped_function(value):
+            self.install_guards(GuardBuilder.TYPE_MATCH)
+            return WrapperUserFunctionVariable(value, "__wrapped__", source=self.source)
         elif is_function_or_wrapper(value):
```
[the remark this answers is not in the document]
Comment (anijain2305) [reply torch/_dynamo/variables/builder.py:909]: Will follow up in the next PR. There is some work remaining in completely removing the need of `unwrap_with_attr_name_if_wrapper`

**Pull request closed without merging**